File: src/main/java/axirassa/services/phone/VoiceNotificationService.java

package axirassa.services.phone;

import java.util.concurrent.Callable;

import lombok.extern.slf4j.Slf4j;

import org.apache.http.client.HttpClient;
import org.apache.http.impl.client.DefaultHttpClient;
import org.hornetq.api.core.client.ClientConsumer;
import org.hornetq.api.core.client.ClientMessage;
import org.hornetq.api.core.client.ClientSession;

import zanoccio.javakit.lambda.Function1;
import axirassa.config.Messaging;
import axirassa.messaging.VoiceRequestMessage;
import axirassa.messaging.util.CommonBackoffStrategies;
import axirassa.messaging.util.InfiniteLoopExceptionSurvivor;
import axirassa.services.Service;
import axirassa.util.MessagingTools;

@Slf4j
public class VoiceNotificationService implements Service {
	private final ClientSession messaging;
	private final HttpClient httpClient;


	public VoiceNotificationService(ClientSession messagingSession) {
		this.messaging = messagingSession;
		this.httpClient = new DefaultHttpClient();
	}


	@Override
	public void execute() throws Throwable {
		messaging.start();

		final ClientConsumer consumer = messaging.createConsumer(Messaging.NOTIFY_VOICE_REQUEST);

		InfiniteLoopExceptionSurvivor executor = new InfiniteLoopExceptionSurvivor(
		        CommonBackoffStrategies.EXPONENTIAL_BACKOFF_MESSAGING(), 
		        new Callable<Object>() {
			        @Override
			        public Object call() throws Exception {
				        ClientMessage message = consumer.receive();
				        System.out.println("####\n####\n####\n####\n####\n####\n");
				        System.out.println("RECIEVED MESSAGE: " + message);

				        VoiceRequestMessage voiceRequest = MessagingTools.fromMessageBytes(VoiceRequestMessage.class,
				                                                                           message);
				        String text = PhoneTemplateFactory.instance.getText(voiceRequest.getTemplate(),
				                                                            PhoneTemplateType.VOICE,
				                                                            voiceRequest.getAttributeMap());

				        SendVoice sender = new SendVoice(voiceRequest.getPhoneNumber(), voiceRequest.getExtension(),
				                text);
				        sender.send(httpClient);

				        message.acknowledge();
				        messaging.commit();
				        return null;
			        }
		        }, new Function1<Boolean, Throwable>() {
			        @Override
			        public Boolean call(Throwable e) throws Throwable {
				        if (e instanceof ClassNotFoundException)
					        log.error("Exception: ", e);
				        else
					        throw e;

				        return null;
			        }
		        });

		executor.execute();
	}
}

File: src/main/java/axirassa/services/phone/VoiceNotificationService.scala

package axirassa.services.phone

import org.apache.http.impl.client.DefaultHttpClient
import org.hornetq.api.core.client.ClientSession

import axirassa.config.Messaging
import axirassa.messaging.VoiceRequestMessage
import axirassa.services.Service
import axirassa.util.AutoSerializingObject

class VoiceNotificationService(messaging : ClientSession) extends Service {
    val httpClient = new DefaultHttpClient()

    override def execute {
        val consumer = messaging.createConsumer(Messaging.NOTIFY_VOICE_REQUEST)
        while (true) {
            try {
                val message = consumer.receive()
                println("####\n####\n####\n####\n####\n####\n")
                println("RECIEVED MESSAGE: "+message)

                val buffer = new Array[Byte](message.getBodyBuffer().readableBytes())
                message.getBodyBuffer().readBytes(buffer)

                val rawobject = AutoSerializingObject.fromBytes(buffer)
                rawobject match {
                    case msg : VoiceRequestMessage => {
                        val text = PhoneTemplateFactory.instance.getText(
                            msg.template,
                            PhoneTemplateType.VOICE,
                            msg.attributeMap)

                        val sender = new SendVoice(msg.phoneNumber, msg.extension, text)
                        sender.send(httpClient)
                    }
                }

                message.acknowledge()
                messaging.commit()
            } catch {
                case e : ClassNotFoundException => e.printStackTrace()
            }
        }
    }
}
